src/Mod/BIM/ArchComponent.py

```diff
@@ -1188,6 +1111,180 @@ def _objectInInternalLinkgroup(self, obj):
                 return True
         return False
 
+class AreaCalculator:
+    """Helper class to compute vertical area, horizontal area, and perimeter length.
+
+    This class encapsulates the logic for calculating the following properties:
+    - **VerticalArea**: The total area of all vertical faces of the object. See the
+      `isFaceVertical` method for the criteria used to determine vertical faces.
+    - **HorizontalArea**: The area of the object's projection onto the XY plane.
+    - **PerimeterLength**: The perimeter of the horizontal area.
+
+    The class provides methods to validate the object's shape, identify vertical and
+    horizontal faces, and compute the required properties.
+    """
+    def __init__(self, obj):
+        self.obj = obj
+
+    def isShapeInvalid(self):
+        """Check if the object's shape is invalid."""
+        return (
+            not self.obj.Shape
+            or self.obj.Shape.isNull()
+            or not self.obj.Shape.isValid()
+            or not self.obj.Shape.Faces
+        )
+
+    def tooManyFaces(self):
+        """Check if the object's shape has too many faces to process."""
+        return len(self.obj.Shape.Faces) > params.get_param_arch("MaxComputeAreas")
+
+    def resetAreas(self):
+        """Reset the area properties of the object to zero. Generally called when
+        there is an error.
+        """
+        for prop in ["VerticalArea", "HorizontalArea", "PerimeterLength"]:
+            setattr(self.obj, prop, 0)
+
+    def isFaceVertical(self, face):
+        """Determine if a face is vertical.
+
+        A face is considered vertical if:
+        - Its normal vector forms an angle close to 90 degrees with the Z-axis.
+        - The face is planar.
+
+        Notes
+        -----
+        The check whether the face is planar means that roof-like (sloped) and domed
+        faces alike will not be counted as vertical faces. It also means though, that
+        vertically-extruded curved edges (for instance from a slab) will also be
+        ignored. See also https://github.com/FreeCAD/FreeCAD/issues/14687.
+        """
+        from Part import OCCError
+        try:
+            angle = face.normalAt(0, 0).getAngle(FreeCAD.Vector(0, 0, 1))
+            return 1.57 < angle < 1.571 and face.Surface.isPlanar()
```

Comment:
```suggestion
            return 1.57 < angle < 1.571 and face.Surface.isPlanar() and face.project(FreeCAD.Vector(0, 0, 1)).Area == 0
```

@paullee0 is this what you were thinking of to classify curved extruded edges correctly as vertical surfaces? Do we need the `isPlanar()` check if we check whether the projected area is 0?

Reply:
Two observations here:

1. No need to test isPlanar - the Surface of vertically extrude curved is not planar, right?  But this should be 'vertical' face.
2. I have tests on face. project(vector), and read the tips from API doc / python console, this seem always return 0 no matter if a face is horizontal or vertical

![image](https://github.com/user-attachments/assets/457fce79-7e2f-41e9-a9a0-106401521bdc)

Reply:
> No need to test isPlanar - the Surface of vertically extrude curved is not planar, right? But this should be 'vertical' face.

That was my thought as well. `isPlanar()` was only added for Arch Space on https://github.com/FreeCAD/FreeCAD/pull/14761. But checking the projection's area is 0 covers both cases of straight and curved projected edges, so it's probably better.

> I have tests on face. project(vector), and read the tips from API doc / python console, this seem always return 0 no matter if a face is horizontal or vertical (tested 0,0,1; 0,1,0; 1,0,0)

I just added this suggestion here for the conversation. I'm not sure if the code line for `TechDraw.project()` here is correct. https://gist.github.com/furgo16/a208ced7386d803e7d3ddc1a81542f20 => TechDraw returns a list with 4 items. Each item is a shape with edges. What I've seen in other places of the code is that:

1. The first item on the list is picked (hard edges shape)
2. Its edges are picked
3. Those edges are converted into wires when they've got coincident points
4. The resulting wires are converted to faces

Here are two ways of doing it (I need to test this):

```python
# Simplest approach, should work for testing areas with 0 value
projectedFace = Part.Face(DraftGeomUtils.findWires(TechDraw.project(f,FreeCAD.Vector(0,0,1))[0].Edges))

# Alternative if we need to account for holes
projectedFace = Part.makeFace(DraftGeomUtils.findWires(TechDraw.project(f,FreeCAD.Vector(0,0,1))[0].Edges), "Part::FaceMakerCheese")

# Then we can check projectedFace.Area 
```

Reply:
I'm also thinking that if there is a tolerance in checking if the normal is 90°, there should also be some tolerance when checking if the area is 0.

Reply:
I've updated the code to do the zero area check only, but it needs to be tested.